socket.c

```diff
@@ -249,18 +249,29 @@ static struct data *construct(const char *spec, int mode) {
   if (data) {
     data->sock = -1;
     data->client_sock = -1;
-
     data->mode = mode;
+
+    char *bracket = strchr(spec, ']');
     char *colon = strchr(spec, ':');
+    char *tail;
+
+    if (bracket) {
+      spec++;
+      tail = (char *)calloc(256, sizeof(char));
```

Comment:
**Every** `malloc` in a C program must be complemented by a `free`. But here you don't even need any additional memory allocation: you should simply pass the right arguments to `strncpy(data->host,`` call.